### ISSUE TITLE
legend ncol causes R crash "invalid color name"

### ISSUE THREAD

**Issue by jmw86069** [jokergoo/ComplexHeatmap#487]
First, I *love* this package! Really amazing, exciting, and thank you so much for your great work!

I am using ComplexHeatmap version `2.3.4`, on `R-3.6.1`.

After some substantial debugging, I found the cause in this minimal test case:

```r
ComplexHeatmap::Heatmap(letters[1:5],
   heatmap_legend_param=list(ncol=4),
   col=rainbow(5))
```

The error message:
```
Error in grid.Call.graphics(C_rect, x$x, x$y, x$width, x$height, resolveHJust(x$just,  :
  invalid color name '�V'
```

The workaround is to use a different `ncol`:

```r
ComplexHeatmap::Heatmap(letters[1:5],
   heatmap_legend_param=list(ncol=3),
   col=rainbow(5))
```

It seems dependent upon the number of items in the legend. For example, with 5 elements in the legend, `ncol=1`, `ncol=2`, `ncol=3`, `ncol=5` all work fine, but `ncol=4` causes an error. It depends on the legend, so I suspect you know where to look. :)

The worst part -- it crashes R*! (Actually it only crashes R on certain systems -- not my Macbook, but it crashes R on a linux server.) Full R crash is pretty rare, and very painful, and challenging for debugging. No worries, it's all the price for being on the cutting edge!

Thank you for any help!

**Comment by jokergoo** [comment]
Hi, here you have 5 elements and you set ncol to 4, which means there are two rows for putting these legends. The default order of putting the elements are by columns, which means actually number of column are only three. I think this inconsistency caused the error. If you add `by_row = TRUE` in the function, you will get what you expect.

```
     [,1] [,2] [,3] [,4]
[1,] "1"  "3"  "5"  ""  
[2,] "2"  "4"  ""   ""
```

For the problem of crashing R, I think it is due to the low-level grid function which accepts invalid values. I will validate user's input and catch this kind of errors.

**Comment by jokergoo** [comment]
Now I have updated the package. It will print messages if `ncol` is not proper.

```
> Heatmap(letters[1:5],
+    heatmap_legend_param=list(ncol=4),
+    col=rainbow(5))
No legend element is put in the last 1 column under `ncol = 4`, maybe
you should set `by_row = TRUE`? Reset `ncol` to 3.
No legend element is put in the last 1 column under `ncol = 4`, maybe
you should set `by_row = TRUE`? Reset `ncol` to 3.
```

**Comment by jmw86069** [comment]
Ahhhhh that makes sense now! Mentally I was doing `"byrow=TRUE"` or `"by_row=TRUE"` even though I saw it was filling by column (which is fine). But it didn't quite click about there being an empty column.

I tested and confirm the newest Github version works as you described! Thanks so much!

Meanwhile, I think my code would be better using `nrow` instead of `ncol` for discrete legends. In your bookdown site section *5.2 Discrete Legends* I see you also explained `nrow` but I missed it the first time.

PS - also very excited to try the new R-shiny features. Wow.
